kernel/kernel-api/src/main/java/io/delta/kernel/internal/tablefeatures/TableFeatures.java

```diff
@@ -414,260 +415,159 @@ public static Tuple2<Integer, Integer> minimumRequiredVersions(Set<TableFeature>
-   * @return true if the "domainMetadata" feature is supported, false otherwise
-   */
   public static boolean isDomainMetadataSupported(Protocol protocol) {
-    return isWriterFeatureSupported(protocol, DOMAIN_METADATA_FEATURE_NAME);
+    return protocol
+        .getImplicitlyAndExplicitlySupportedFeatures()
+        .contains(DOMAIN_METADATA_W_FEATURE);
   }
 
+  /////////////////////////////////////////////////////////////////////////////////
+  /// Private methods                                                           ///
+  /////////////////////////////////////////////////////////////////////////////////
   /**
-   * Check if the table protocol supports the "rowTracking" writer feature.
-   *
-   * @param protocol the protocol to check
-   * @return true if the protocol supports row tracking, false otherwise
+   * Extracts all new table features (and their dependency features) that are supported by the given
+   * metadata. A feature is considered new if it is not already supported by the current protocol.
    */
-  public static boolean isRowTrackingSupported(Protocol protocol) {
-    return isWriterFeatureSupported(protocol, ROW_TRACKING_FEATURE_NAME);
-  }
+  private static Set<TableFeature> extractAutomaticallyEnabledNewFeatures(
+      Metadata newMetadata, Protocol currentProtocol) {
+    Set<TableFeature> protocolSupportedFeatures =
+        currentProtocol.getImplicitlyAndExplicitlySupportedFeatures();
 
-  /**
-   * Get the minimum reader version required for a feature.
-   *
-   * @param feature the feature
-   * @return the minimum reader version required for the feature
-   */
-  private static int getMinReaderVersion(String feature) {
-    switch (feature) {
-      case "inCommitTimestamp":
-        return 3;
-      default:
-        return 1;
-    }
-  }
+    Set<TableFeature> metadataEnabledFeatures =
+        TableFeatures.TABLE_FEATURES.stream()
+            .filter(f -> f instanceof FeatureAutoEnabledByMetadata)
+            .filter(
+                f ->
+                    ((FeatureAutoEnabledByMetadata) f)
+                        .metadataRequiresFeatureToBeEnabled(currentProtocol, newMetadata))
+            .collect(toSet());
 
-  /**
-   * Get the minimum writer version required for a feature.
-   *
-   * @param feature the feature
-   * @return the minimum writer version required for the feature
-   */
-  private static int getMinWriterVersion(String feature) {
-    switch (feature) {
-      case "domainMetadata": // fall through
-      case "inCommitTimestamp":
-        return 7;
-      default:
-        return 2;
-    }
+    Set<TableFeature> combinedFeatures = new HashSet<>(protocolSupportedFeatures);
+    combinedFeatures.addAll(metadataEnabledFeatures);
+
+    // Each feature may have dependencies that are not yet enabled in the protocol.
+    Set<TableFeature> newFeatures = getDependencyFeatures(combinedFeatures);
+    newFeatures.removeAll(protocolSupportedFeatures);
```

Comment:
why do we need to do `getDependencyFeatures` on `combinedFeatures`? And not just on the `newFeatures`?

Reply:
I think the newFeatures should be good enough.